File: backend-kotlin/app/api/src/main/kotlin/com/wallet/clover/api/service/FcmService.kt
package com.wallet.clover.api.service

import com.google.firebase.FirebaseApp
import com.google.firebase.messaging.FirebaseMessaging
import com.google.firebase.messaging.Message
import com.google.firebase.messaging.Notification
import com.wallet.clover.api.repository.user.UserRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class FcmService(
    private val userRepository: UserRepository
) {
    private val logger = LoggerFactory.getLogger(javaClass)

    @Transactional
    suspend fun registerToken(ssoQualifier: String, token: String) {
        val user = userRepository.findBySsoQualifier(ssoQualifier)
        if (user != null) {
            userRepository.save(user.copy(fcmToken = token))
            logger.info("사용자 ${user.id}의 FCM 토큰 등록: $token")
        }
    }

    /**
     * 특정 사용자에게 푸시 알림 전송
     * @param token FCM 디바이스 토큰
     * @param title 알림 제목
     * @param body 알림 본문
     */
    suspend fun sendToUser(token: String, title: String, body: String) {
        if (FirebaseApp.getApps().isEmpty()) {
            logger.warn("Firebase가 초기화되지 않았습니다. FCM 알림을 건너뜁니다.")
            return
        }

        withContext(Dispatchers.IO) {
            try {
                val notification = Notification.builder()
                    .setTitle(title)
                    .setBody(body)
                    .build()

                val message = Message.builder()
                    .setToken(token)
                    .setNotification(notification)
                    .build()

                val response = FirebaseMessaging.getInstance().send(message)
                logger.info("사용자에게 메시지 전송 성공: {}", response)
            } catch (e: Exception) {
                logger.error("토큰으로 FCM 알림 전송 오류: $token", e)
                // 에러가 발생해도 계속 진행 (non-blocking)
            }
        }
    }

    /**
     * 로또 당첨 알림 전송
     */
    suspend fun sendWinningNotification(token: String, rank: String, numbers: List<Int>, amount: Long? = null) {
        val title = "🎉 로또 당첨!"
        val amountText = amount?.let { " (당첨금: ${it}원)" } ?: ""
        val body = "$rank 당첨!$amountText 번호: ${numbers.sorted().joinToString(", ")}"
        
        if (FirebaseApp.getApps().isEmpty()) {
            logger.warn("Firebase가 초기화되지 않았습니다. FCM 당첨 알림을 건너뜁니다.")
            return
        }

        withContext(Dispatchers.IO) {
            try {
                val notification = Notification.builder()
                    .setTitle(title)
                    .setBody(body)
                    .build()

                val message = Message.builder()
                    .setToken(token)
                    .setNotification(notification)
                    .putData("type", "WINNING")         // 알림 타입
                    .putData("screen", "history")       // 이동할 화면
                    .putData("rank", rank)              // 당첨 등급
                    .build()

                FirebaseMessaging.getInstance().send(message)
                logger.info("사용자에게 당첨 알림 전송: $title - $body")
            } catch (e: Exception) {
                logger.error("당첨 알림 전송 오류", e)
            }
        }
    }

    /**
     * 다수의 사용자에게 동일한 알림 브로드캐스트
     * @param tokens FCM 디바이스 토큰 리스트
     * @param title 알림 제목
     * @param body 알림 본문
     */
    suspend fun sendBroadcastNotification(tokens: List<String>, title: String, body: String) {
        if (FirebaseApp.getApps().isEmpty()) {
            logger.warn("Firebase가 초기화되지 않았습니다. FCM 브로드캐스트 알림을 건너뜁니다.")
            return
        }

        withContext(Dispatchers.IO) {
            var successCount = 0
            var failureCount = 0

            tokens.forEach { token ->
                try {
                    val notification = Notification.builder()
                        .setTitle(title)
                        .setBody(body)
                        .build()

                    val message = Message.builder()
                        .setToken(token)
                        .setNotification(notification)
                        .putData("type", "DRAW_REMINDER")      // 알림 타입
                        .putData("screen", "number_generation") // 이동할 화면
                        .build()

                    FirebaseMessaging.getInstance().send(message)
                    successCount++
                } catch (e: Exception) {
                    logger.error("토큰으로 FCM 전송 오류: $token", e)
                    failureCount++
                }
            }

            logger.info("브로드캐스트 알림 전송: 성공 $successCount, 실패 $failureCount")
        }
    }
}

File: backend-node/src/notification/fcm.service.ts
import { Injectable, Logger, OnModuleInit } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import * as admin from 'firebase-admin';
import { ConfigService } from '@nestjs/config';

/**
 * FCM(Firebase Cloud Messaging)을 통한 푸시 알림 발송을 담당하는 서비스입니다.
 * Kotlin FcmService 로직을 이식함.
 */
@Injectable()
export class FcmService implements OnModuleInit {
  private readonly logger = new Logger(FcmService.name);
  private isFirebaseInitialized = false;

  constructor(
    private readonly prisma: PrismaService,
    private readonly configService: ConfigService,
  ) {}

  onModuleInit() {
    this.initializeFirebase();
  }

  private initializeFirebase() {
    try {
      const serviceAccountPath = this.configService.get<string>('FIREBASE_KEY_PATH');
      
      if (serviceAccountPath && !admin.apps.length) {
        admin.initializeApp({
          credential: admin.credential.cert(serviceAccountPath),
        });
        this.isFirebaseInitialized = true;
        this.logger.log('Firebase Admin SDK 초기화 성공');
      } else {
        this.logger.warn('FIREBASE_KEY_PATH가 설정되지 않았습니다. FCM 기능을 사용할 수 없습니다.');
      }
    } catch (error) {
      this.logger.error('Firebase Admin SDK 초기화 실패', error.stack);
    }
  }

  /**
   * 사용자의 FCM 토큰을 ID 기반으로 등록/업데이트합니다.
   */
  async registerTokenById(userId: bigint, token: string) {
    await this.prisma.user.update({
      where: { id: userId },
      data: { fcmToken: token },
    });
    this.logger.log(`사용자 ${userId}의 FCM 토큰 등록 완료 (ID 기반)`);
  }

  /**
   * 사용자의 FCM 토큰을 등록/업데이트합니다.
   * @param ssoQualifier 사용자 SSO 식별자
   * @param token FCM 토큰
   */
  async registerToken(ssoQualifier: string, token: string) {
    const user = await this.prisma.user.findUnique({
      where: { ssoQualifier },
    });

    if (user) {
      await this.prisma.user.update({
        where: { id: user.id },
        data: { fcmToken: token },
      });
      this.logger.log(`사용자 ${user.id}의 FCM 토큰 등록 완료`);
    }
  }

  /**
   * 특정 사용자에게 당첨 알림을 전송합니다.
   */
  async sendWinningNotification(
    token: string,
    rank: string,
    numbers: number[],
    amount?: bigint,
  ) {
    const title = '🎉 로또 당첨!';
    const amountText = amount ? ` (당첨금: ${amount.toLocaleString()}원)` : '';
    const body = `${rank} 당첨!${amountText} 번호: ${numbers.sort((a, b) => a - b).join(', ')}`;

    if (!this.isFirebaseInitialized) {
      this.logger.warn(`FCM 미초기화로 알림 전송 스킵: ${body}`);
      return;
    }

    try {
      await admin.messaging().send({
        token,
        notification: { title, body },
        data: {
          type: 'WINNING',
          screen: 'history',
          rank,
        },
      });
      this.logger.log(`당첨 알림 전송 성공: ${token}`);
    } catch (error) {
      this.logger.error(`당첨 알림 전송 실패: ${token}`, error.stack);
    }
  }

  /**
   * 다수 사용자에게 알림을 브로드캐스트합니다.
   */
  async sendBroadcastNotification(
    tokens: string[],
    title: string,
    body: string,
  ) {
    if (!this.isFirebaseInitialized || tokens.length === 0) return;

    const messages = tokens.map((token) => ({
      token,
      notification: { title, body },
      data: { type: 'DRAW_REMINDER' },
    }));

    try {
      const response = await admin.messaging().sendEach(messages);
      this.logger.log(
        `브로드캐스트 전송 완료: 성공 ${response.successCount}, 실패 ${response.failureCount}`,
      );
    } catch (error) {
      this.logger.error('브로드캐스트 전송 실패', error.stack);
    }
  }
}
